File: BST Maximum Difference[23-March-2023]/Solution.cpp
//				>> C++ CODE <<


class Solution {
public:
    int find(Node* root, int target, Node** targetNode) {
        if (root == NULL) return INT_MIN;
        if (root->data == target) {
            *targetNode = root;
            return target;
        }
        
        if (target > root->data) {
            return root->data + find(root->right, target, targetNode);
        } else {
            return root->data + find(root->left, target, targetNode);
        }
    }
    
    int solve(Node* root) {
        if (root == NULL) return INT_MAX / 2;
        if (root->left == NULL && root->right == NULL) return root->data;
        
        int left = root->data + solve(root->left);
        int right = root->data + solve(root->right);
        
        return min(left, right);
    }
    
    int maxDifferenceBST(Node* root, int target) {
        Node* targetNode = NULL;
        int rootToTarget = find(root, target, &targetNode);
        
        if (targetNode == NULL) return -1;
        
        int targetToLeaf = solve(targetNode);
        
        return rootToTarget - targetToLeaf;
    }
};

File: BST Maximum Difference[23-March-2023]/Solution.java
//			>> JAVA CODE <<

class Solution
{
    public static int find(Node root,int target,Node[] targetNode)
    {
        if(root==null) return Integer.MIN_VALUE;
        if(root.data==target){
          targetNode[0]=root;  
          return target;
        }
        int left=find(root.left,target,targetNode);
        int right=find(root.right,target,targetNode);
        
        if(left==Integer.MIN_VALUE) 
        {
            if(right==Integer.MIN_VALUE) return Integer.MIN_VALUE;
            else return root.data+right;
        }
        else
        {
            return root.data+left;  
        }
    }
    public static int solve(Node root)
    {
        if(root==null) return Integer.MAX_VALUE/2;
        if(root.left==null && root.right==null) return root.data;
        
        int left=root.data+solve(root.left);
        int right=root.data+solve(root.right);
        
            return Math.min(left,right);
    }
    public static int maxDifferenceBST(Node root,int target)
    {
        Node[] targetNode={null};
        int rootToTarget=find(root,target,targetNode);
        
        if(targetNode[0]==null) return -1;
        
        
        int targetToLeaf=solve(targetNode[0]);
        
            return rootToTarget-targetToLeaf;
    }
}